### DISCUSSION TITLE
TypeScript Conditional prop rendering throws errors

### DISCUSSION THREAD

**Discussion by thiagodesa26** [storybookjs/storybook#18572] [Help]
**Background:**
The Button component has conditional props. When buttonType === "edit", then children is not allowed. When the buttonType !== "edit", then children is required.

Typescript and react works fine with the conditional props. No errors and the functionality is correct. Only with storybook I can't use it.

**Code to validate:**

**Button.tsx:**

```
import React, { ReactChild } from "react";
import EditIco from "assets/icons/ico-edit.svg";

type ButtonType = "slim" | "normal" | "large" | "xlarge" | "back" | "edit";

type ButtonProps =
  | {
      onClick?: () => void;
    } & (
      | {
          buttonType: "edit";
          children?: never;
        }
      | {
          buttonType: Exclude<ButtonType, "edit">;
          children: string | ReactChild;
        }
    );

const Button = ({
  children,
  onClick = () => {},
  buttonType,
}: ButtonProps) => {
  const isEditButton = buttonType === "edit";

  return (
    <button
      type="button"
      className={`${styles["buttonType-" + buttonType]}`}
      onClick={onClick}
    >
      {isEditButton ? <img src={EditIco} alt="Edit" /> : children}
    </button>
  );
};

export default Button;
```
**Button.stories.tsx:**

```
import { ComponentStory, ComponentMeta } from "@storybook/react";
import Button from "./Button";

export default {
  title: "Atoms/Button",
  component: Button,
  argTypes: {},
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => (
  <Button {...args}>{args.children}</Button>
);

export const Large = Template.bind({});
Large.args = {
  buttonType: "large",
  children: "Large",
};

```
Storybook fails as it loads the edit prop expecting children not to be passed. Even if there is no binding export using args with `buttonType: "edit"`.

![image](https://user-images.githubusercontent.com/24533166/175807322-2fb6615e-0204-4266-bb0e-49b21c55b119.png)

Question:

Which would be the correct way to set the template so it would bind all buttonType variants except "edit" ?

**Comment by will-stone** [comment]
Hi @thiagodesa26 did you happen to find a solution for this?
